proxy/src/keystore.rs

```diff
@@ -0,0 +1,211 @@
+//! Storage of secret keys.
+
+use std::convert::Infallible;
+use std::fmt;
+
+use secstr::{SecStr, SecUtf8};
+
+use librad::keys;
+use librad::paths;
+use radicle_keystore::{
+    crypto::{Pwhash, SecretBoxError},
+    file, FileStorage, Keystore, SecretKeyExt,
+};
+use radicle_registry_client::{ed25519, CryptoError, CryptoPair};
+
+const LIBRAD_KEY: &str = "librad.key";
+const REGISTRY_KEY: &str = "registry.key";
+
+/// Storage for putting and getting the necessary cryptographic keys.
+pub struct Keystorage {
+    librad_store: LibradStore,
+    registry_store: RegistryStore,
+}
+
+impl Keystorage {
+    /// Create a new `CocoStore` for storing your [`keys::SecretKey`].
+    #[must_use = "must use CocoStore to put/get a key"]
+    pub fn new(paths: &paths::Paths, pw: SecUtf8) -> Result<Keystorage, Error> {
+        let path = paths.keys_dir();
```

Comment:
It feels invasive to use the librad owned paths here, esp. as we control the name of the keys and therefore can't assume overlap with other librad usages outside of the application. We have our own data tree which we use for our session store already and should expand with a keys sub directory:

https://github.com/radicle-dev/radicle-upstream/blob/0ada292486946746fe5eef9046f7b280ae70d899/proxy/src/main.rs#L105

Reply:
The problem is that upstream can't own these key paths either -- otherwise all other implementations (eg. CLIs) will have to hijack the `upstream` namespace for their XDG needs.

I'm not entirely happy with the fact that `librad`, being a library, defines a fixed config location, but I also don't see a better solution -- those locations are, in fact, global. Iow, I don't know a better way than to just acknowledge the existence of the trinity, and standardise on three different namespaces:

    xyz.radicle.upstream
    xyz.radicle.link
    xyz.radicle.registry-client

It would, however, be sufficient if `librad` would implement `TryFrom<ProjectDirs> for Paths` instead of hardcoding the namespace. The namespacing scheme just needs to be documented, so alternative implementations can share it.

Reply:
> The problem is that upstream can't own these key paths either -- otherwise all other implementations (eg. CLIs) will have to hijack the `upstream` namespace for their XDG needs.

This makes a lot of sense, if we want that then librad should dictate/own the entire location of the key including the file name, or am I missing something?

Reply:
Should this mean that registry should also dictate where their key lives?

Reply:
> Should this mean that registry should also dictate where their key lives?

Not as I'm aware of. Maybe @radicle-dev/registry can shed some light on how keys are usually handling in other tools like the cli for registry.

Reply:
Well, I'd say that it's a good idea to do the same thing. Since the whole concept is key safety, where we can read the encrypted key from storage at the point where it's necessary, and then drop the key again so that it doesn't live in memory.

But ya, would be nice to get the input from registry on this. I can follow up with any changes if they're needed.